### PR TITLE
Peer Code Review

### PR DESCRIPTION
**Pull request by andela-mkamau**
I would like you to have a look at my implementation of tests and functionality of checkpoint 1.
Please, make comments, suggestions and criticisms.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    │ Snoe          │
+    ├───────────────┤
+    │ Oculus        │
+    ╘═══════════════╛
+    ```
+    
+See the video below to get a compelete picture of how to use Office
+Room Allocator:
+
+
```
Comment (Stephen-Njoroge) [inline README.md:259]: Nice Work on the readMe it is pretty detailed and structured, also make sure to merge all your changes to master at the end since `git clone https://github.com/andela-mkamau/Office-Space-Allocation.git` clones an empty repo currently.

File: office_space_allocation/amity.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    def has_room(self, room):
+        """
+        Checks if Room `room` already exists in Amity
+        :return: `True` if room exist, otherwise `False`
+        """
+        for r in self.all_rooms:
+            if r == room:
+                return True
+        return False
+
```
Comment (Stephen-Njoroge) [inline office_space_allocation/amity.py:43]: How about `if room in self.all_rooms:` It is shorter and more pythonic than 
`for r in self.all_rooms:
            if r == room:
               return True
            return False`

File: office_space_allocation/amity.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    def find_room(self, name):
+        """
+        Find a ```Room``` object using name
+        :param name: ```str``` name of the Room to look for
+        :return: ```Room``` object if found
+        """
```
Comment (Stephen-Njoroge) [inline office_space_allocation/amity.py:49]: Nice work documenting your code throughout

File: office_space_allocation/amity.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        name = name.lower()
+        for r in self.all_rooms:
+            if name == r.get_name().lower():
+                return r
+        raise ValueError(name + " not found!")
+
```
Comment (Stephen-Njoroge) [inline office_space_allocation/amity.py:55]: You can also shorten and optimise your code by this by `if name in self.all_rooms`. You can optimise your code to append names to the list in either lowercase i.e .lower(), uppercase or .title() for the first letter to be capital, so that names in rooms can have a uniform format rather than converting them when applying functions.

File: office_space_allocation/amity.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        """
+        Allocates a room to ```Person``` person.
+
+        This is done randomly.
+        :param person: ```Person``` to allocate room
+        :return: room : ```Room``` allocated , if possible
+        """
+        if isinstance(person, Staff):
+            # allocate only offices
+            offices = [of for of in self.all_rooms if isinstance(of, Office)]
```
Comment (Stephen-Njoroge) [inline office_space_allocation/amity.py]: `office for office in self.all_rooms` It is good to name variables appropriately so that your code is self-documenting.

File: office_space_allocation/amity.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            if len(offices) < 1:
```
Comment (Stephen-Njoroge) [inline office_space_allocation/amity.py]: Why not 
 ```
   if  len(offices):
        do  all your staff
    raise IndexError
```

File: office_space_allocation/amity.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                                 "allocations")
+            else:
+                rm = random.choice(offices)
+
+                rm.add_person(person)
+                self.allocated_rooms.append(rm)
+                return rm
+        elif isinstance(person, Fellow):
+            # raise exception if there are no rooms
+            if len(self.all_rooms) < 1:
```
Comment (Stephen-Njoroge) [inline office_space_allocation/amity.py]:
```
if len(self.all_rooms):
    do all your staff
raise error
```

File: office_space_allocation/amity.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+    def find_person(self, name):
+        """
+        Finds a Person using name.
+        Name can be the full name, first name, last name or part of either
+        :param name:
+        :return: a tuple with Person(s) objects found in the list if persons
+        """
+        name = name.strip()
+        res = [p for p in self.all_persons if name.title() in p.get_full_name()]
```
Comment (Stephen-Njoroge) [inline office_space_allocation/amity.py:117]: Avoid using such variable names such as res and for p .... They make your code hard to follow when you are working in teams or in the case of another developer having to work with your code.

File: office_space_allocation/amity.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+        try:
+            room = self.find_room(room_name)
+        except ValueError:
+            raise
+
+        if room.has_person(person):
+            raise Exception("Cannot reallocate person to same room.\n {} is already in {}".format(person_name,
+                                                                                                  room_name))
+
```
Comment (Stephen-Njoroge) [inline office_space_allocation/amity.py:153]: What is happening here...
I don't know why the code has gone so many lines.

File: office_space_allocation/staff.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
+from office_space_allocation.person import Person
+
+
+class Staff(Person):
+    def __eq__(self, other_person):
+        """
+        Tests for equality between this Person and the other_person
```
Comment (Stephen-Njoroge) [inline office_space_allocation/staff.py:7]: More clariffication here like what is other_person.

File: office_space_allocation/amity.py (hunk 2 above)
Comment (andela-aomondi) [inline office_space_allocation/amity.py]: You could rewrite as `if room on self.all_rooms`

File: osa.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+     osa.py [-i | --interactive]
+     osa.py quit
+
+"""
+
+import cmd
+import sys
+
+from docopt import docopt, DocoptExit
+from colorama import Fore, Style
```
Comment (amos-o) [inline osa.py:25]: Reorder. Coloroma should come before docopt

File: README.md [hunk 12]
```diff
@@ -65,30 +65,193 @@ Office Space Allocator is run in an interactive mode. Functionality is
...
 
-**NOTE: You cannot create rooms with the same name**. Am sure you do want to
+    ```
+    (amity >>> ) create_room livingspace Kitchen Python Php
+    Successfully created livingspace room: Kitchen
+    Successfully created livingspace room: Python
+    Successfully created livingspace room: Php
+    ```
+    
+    **NOTE: You cannot create rooms with the same name**. Am sure you do want to
```
Comment (amos-o) [inline README.md:90]: This should be `you do not want to`

Comment (andela-mkamau) [reply README.md:90]: Yeah, I don't how I missed that. Thanks

File: osa.py (hunk 11 above)
Comment (andela-mkamau) [reply osa.py:25]: Right. Fixed that already

File: office_space_allocation/main.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+from office_space_allocation import amity, office, staff, livingspace, fellow, utilities, db
+
+main_amity = None
+
+
+def init_amity():
+    """
+    Initialises the Amity class
+    """
+    global main_amity
```
Comment (amos-o) [inline office_space_allocation/main.py:19]: I do not think line 12 is required since on line 19 you are redefining main_amity as a global variable.

Comment (amos-o) [reply office_space_allocation/main.py:19]: Or you could remove line 19 and work with line 12

Comment (andela-mkamau) [reply office_space_allocation/main.py:19]: I think I could actually achieve the functionality I need by only a `main_amity = amity.Amity()` . The reason I introduced the function `init_amity()` was to prevent setting the variable `main_amity` to an instance of `amity.Amity()` whenever the module is imported. In other words, I wanted it set explicitly via  the function `init_amity()`.
